### PR TITLE
Throw an Exception instead of an error on Cache::write() fail. See original #16932

### PR DESCRIPTION
**Pull request by fr3nch13**
Closing #16932 and opening this one so the branches line up.

### CODE REVIEW

**Comment by fr3nch13** [comment]
I guess my phpstan/phpunit coverage isn't as good as I thought. I'm not sure what needs to be done so that phpstan will pass, and codecov will pas as well. Anyone have any ideas/pointers?

**Comment by jamisonbryant** [comment]
For the stan fail, could this be the issue?

```
$this->expectException('Cake\Cache\InvalidArgumentException');
```

It looks like the path is supposed to be `Cake\Cache\Exception\InvalidArgumentException`, no?

File: tests/TestCase/Cache/InvalidArgumentExceptionTest.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * InvalidArgumentExceptionTest class
+ */
+class InvalidArgumentExceptionTest extends TestCase
+{
+    /**
+     * tests That the old/deprecated class is properly aliased to the new one.
+     */
+    public function testThrowingOldException(): void
+    {
+        $this->expectException(\Cake\Cache\InvalidArgumentException::class);
```
Comment (othercorey) [inline tests/TestCase/Cache/InvalidArgumentExceptionTest.php]: Use a `use` statement instead of the full path.

Comment (dereuromark) [reply tests/TestCase/Cache/InvalidArgumentExceptionTest.php]: Shouldn't cs fix this?

Comment (othercorey) [reply tests/TestCase/Cache/InvalidArgumentExceptionTest.php]: I think it is failing, just trying to be helpful.

Comment (dereuromark) [reply tests/TestCase/Cache/InvalidArgumentExceptionTest.php]: No, thing is CI is green on that part
We might need a sniffer rule here

Comment (ADmad) [reply tests/TestCase/Cache/InvalidArgumentExceptionTest.php]: The sniff for this exists in the our v5 ruleset, not v4.

Comment (fr3nch13) [reply tests/TestCase/Cache/InvalidArgumentExceptionTest.php]: It was passing, but codecov was failing as it was lowering coverage, that's why I added it. I can remove it.

Comment (fr3nch13) [reply tests/TestCase/Cache/InvalidArgumentExceptionTest.php]: I removed it, and codecov dropped it's coverage percentage. Will that sniff in 5.x cover it?

File: tests/TestCase/View/StringTemplateTest.php [hunk 2]
```diff
@@ -80,6 +80,27 @@ public function testAddWithInvalidTemplate(): void
         ];
         $this->expectException(InvalidArgumentException::class);
         $this->template->add($templates);
+
+        $templates = [
+            'link' => '<a href="{{url}}">{{text}}</a>',
+            'invalid' => false,
+        ];
+        $this->expectException(InvalidArgumentException::class);
+        $this->template->add($templates);
```
Comment (ADmad) [inline tests/TestCase/View/StringTemplateTest.php]: How are these tests related to your changes?

Comment (fr3nch13) [reply tests/TestCase/View/StringTemplateTest.php]: It was related to the other pull request. It checks for more then just a null template. I removed it from here.

File: src/Cache/Exception/CacheWriteException.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         4.4.12
```
Comment (ADmad) [inline src/Cache/Exception/CacheWriteException.php]:
```suggestion
 * @since         4.5.0
```

Comment (fr3nch13) [reply src/Cache/Exception/CacheWriteException.php]: Should the sniff be brought over as well for the exception?

**Comment by fr3nch13** [comment]
Is there anything else you all need for this pull?

**Comment by fr3nch13** [comment]
So this should be passing with the exception of codecov/patch. What's needed to make it happy?

File: tests/TestCase/Cache/Engine/MemcachedEngineTest.php [hunk 4]
```diff
@@ -18,9 +18,9 @@
 
 use Cake\Cache\Cache;
 use Cake\Cache\Engine\MemcachedEngine;
+use Cake\Cache\Exception\InvalidArgumentException;
 use Cake\TestSuite\TestCase;
 use DateInterval;
-use InvalidArgumentException;
```
Comment (othercorey) [inline tests/TestCase/Cache/Engine/MemcachedEngineTest.php:23]: Looks like this test is using \InvalidArgumentException and shouldn't be changed.

File: VERSION.txt [hunk 5]
```diff
@@ -16,4 +16,4 @@
 // @license       https://opensource.org/licenses/mit-license.php MIT License
 // +--------------------------------------------------------------------------------------------+ //
 ////////////////////////////////////////////////////////////////////////////////////////////////////
-4.5.0-dev
+4.5.0-dev
```
Comment (othercorey) [inline VERSION.txt:19]: Can you revert this file?
